src/couch_stats/src/csrt_util.erl

```diff
@@ -0,0 +1,485 @@
+% License for the specific language governing permissions and limitations under
+% the License.
+
+-module(csrt_util).
+
+-export([
+    is_enabled/0,
+    is_enabled_init_p/0,
+    get_pid_ref/0,
+    get_pid_ref/1,
+    set_pid_ref/1,
+    should_track_init_p/2,
+    tnow/0,
+    tutc/0,
+    tutc/1
+]).
+
+%% JSON Conversion API
+-export([
+    convert_type/1,
+    convert_pidref/1,
+    convert_pid/1,
+    convert_ref/1,
+    to_json/1
+]).
+
+%% Delta API
+-export([
+    add_delta/2,
+    extract_delta/1,
+    get_delta/1,
+    get_delta_a/0,
+    get_delta_zero/0,
+    maybe_add_delta/1,
+    maybe_add_delta/2,
+    make_delta/1,
+    make_dt/2,
+    make_dt/3,
+    rctx_delta/2,
+    set_delta_a/1,
+    set_delta_zero/1
+]).
+
+%% Extra niceties and testing facilities
+-export([
+    set_fabric_init_p/2,
+    set_fabric_init_p/3,
+    map_to_rctx/1,
+    field/2
+]).
+
+-include_lib("couch_stats_resource_tracker.hrl").
+
+-spec is_enabled() -> boolean().
+is_enabled() ->
+    config:get_boolean(?CSRT, "enabled", true).
+
+-spec is_enabled_init_p() -> boolean().
+is_enabled_init_p() ->
+    config:get_boolean(?CSRT_INIT_P, "enabled", true).
+
+-spec should_track_init_p(Mod :: atom(), Func :: atom()) -> boolean().
+should_track_init_p(fabric_rpc, Func) ->
+    is_enabled_init_p() andalso config:get_boolean(?CSRT_INIT_P, fabric_conf_key(Func), false);
+should_track_init_p(_Mod, _Func) ->
+    false.
+
+%% Monotnonic time now in native format using time forward only event tracking
+-spec tnow() -> integer().
+tnow() ->
+    erlang:monotonic_time().
+
+%% Get current system time in UTC RFC 3339 format
+-spec tutc() -> calendar:rfc3339_string().
+tutc() ->
+    tutc(tnow()).
+
+%% Convert a integer system time in milliseconds into UTC RFC 3339 format
+-spec tutc(Time :: integer()) -> calendar:rfc3339_string().
+tutc(Time0) when is_integer(Time0) ->
+    Unit = millisecond,
+    Time1 = Time0 + erlang:time_offset(),
+    Time = erlang:convert_time_unit(Time1, native, Unit),
+    calendar:system_time_to_rfc3339(Time, [{unit, Unit}, {offset, "z"}]).
+
+%% Returns dt (delta time) in microseconds
+%% @equiv make_dt(A, B, microsecond)
+-spec make_dt(A, B) -> pos_integer() when
+    A :: integer(),
+    B :: integer().
+make_dt(A, B) ->
+    make_dt(A, B, microsecond).
+
+%% Returns monotonic dt (delta time) in specified time_unit()
+-spec make_dt(A, B, Unit) -> pos_integer() when
+    A :: integer(),
+    B :: integer(),
+    Unit :: erlang:time_unit().
+make_dt(A, A, _Unit) when is_integer(A) ->
+    %% Handle edge case when monotonic_time()'s are equal
+    %% Always return a non zero value so we don't divide by zero
+    %% This always returns 1, independent of unit, as that's the smallest
+    %% possible positive integer value delta.
+    1;
+make_dt(A, B, Unit) when is_integer(A) andalso is_integer(B) andalso B > A ->
+    A1 = erlang:convert_time_unit(A, native, Unit),
+    B1 = erlang:convert_time_unit(B, native, Unit),
```

Comment:
Subtract native first then convert once?

Reply:
Hmm, but if we want natural numbers the conversion order (convert then subtract or subtract then convert) won't help as much, we could still get a 0?

For instance, with B > A (given the guard) in native time units if we convert them to milliseconds we could easily and up with a 0

```
1> f(), {A,B} = {erlang:monotonic_time(), erlang:monotonic_time()}.
{-576460747938413313,-576460747938410997}
2> AMsec = erlang:convert_time_unit(A, native, millisecond).
-576460747939
3> BMsec = erlang:convert_time_unit(B, native, millisecond).
-576460747939
4> BMsec - AMsec.
0
```